overlord/snapstate/handlers.go

```diff
@@ -1362,15 +1461,42 @@ func (m *SnapManager) undoLinkSnap(t *state.Task, _ *tomb.Tomb) error {
 		return err
 	}
 
-	// we need to undo potential changes to current snap configuration (e.g. if modified by post-refresh/install/configure hooks
-	// as part of failed refresh/install) by restoring the configuration of "old current".
+	// we need to undo potential changes to current snap configuration (e.g. if
+	// modified by post-refresh/install/configure hooks as part of failed
+	// refresh/install) by restoring the configuration of "old current".
+	// similarly, we need to re-save the disabled services if there is a
+	// revision for us to go back to, see comment below for full explanation
 	if len(snapst.Sequence) > 0 {
 		if err = config.RestoreRevisionConfig(st, snapsup.InstanceName(), oldCurrent); err != nil {
 			return err
 		}
+
+		// unlock state while we talk to systemd
+		st.Unlock()
+		defer st.Lock()
+
+		// get the currently disabled services and add them to
+		// snapst.LastActiveDisabledServices because if we completed a successful
+		// doLinkSnap (hence we are in the undo handler), then we already disabled
+		// the services and deleted currently existing services from the state
+		// during doLinkSnap, but now we will need that information again when we go
+		// to link the old version to prevent accidental enabling of disabled
+		// services on a failed revert/refresh
+		disabledServices, err := m.queryDisabledServices(newInfo, NewTaskProgressAdapterUnlocked(t))
```

Comment:
shouldn't we store the info as some kind of old* value as we do for other things?  if we are undoing how do we know this information isn't messed up in some way?

Reply:
hmm I guess I went more with the approach of "be very careful when writing to state" so the old variable stuff didn't seem necessary, but I suppose if this information was messed up somehow we should store it that way instead.

Reply:
I'm slightly confused about the old* pattern, as it seems we only use this pattern for specifically `doLinksnap`/`undoLinkSnap` and don't do it for example with `doUnlinkCurrentSnap`/`undoUnlinkCurrentSnap`. 

Should I also use the old* pattern for `doUnlinkCurrentSnap`/`undoUnlinkCurrentSnap`?

Reply:
I'm not sure I understand the question.  It depends which places change and how LastActiveDisabledServices, we need to do that consistently, that's why as I said unit tests and tests about where that happens are important. Afaict LastActiveDisabledServices is its own old wrt (do/undo)UnlinkCurrentSnap, unless I'm missing something.

Reply:
Okay I think I understand your question. 

How I originally had coded it, yes LastActiveDisabledServices was effectively its own old variable, usable for both successful forward ops like install and undo ops. 

In retrospect, I don't think it was necessary to use the old variables in doLinkSnap/undoLinkSnap unless we are worried about state itself getting corrupted. However, I can see how to implement the old variable pattern for (do/undo)UnlinkCurrentSnap, so I can implement it that way to be consistent if you like. Otherwise perhaps I should revert the change doing the old variables for (do/undo)LinkSnap to be the same

Reply:
> However, I can see how to implement the old variable pattern for (do/undo)UnlinkCurrentSnap, so I can implement it that way to be consistent if you like.

undoUnlinkCurrentSnap and undoLinkSnap don't need to use consistent patterns, the reason one uses old* and the other not is they actually deal with very different situations. One doesn't do much except toggling Active and the link (etc). The other crosses revision potentially.

Basically while it's expected for undoLinkSnap to use old*, it's should be needed by undoUnlinkCurrentSnap.

Reply:
As discussed in SU, we will keep old* for undoLinkSnap, we don't need to do this for undoUnlinkCurrentSnap.